substrate/network/src/consensus.rs

```diff
@@ -0,0 +1,203 @@
+	completion: oneshot::Sender<Vec<u8>>,
+}
+
+struct PeerConsensus {
+	candidate_fetch: Option<CandidateRequest>,
+	candidate_available: Option<Hash>,
+}
+
+/// Consensus network protocol handler. Manages statements and candidate requests.
+pub struct Consensus {
+	peers: HashMap<PeerId, PeerConsensus>,
+	our_candidate: Option<(Hash, Vec<u8>)>,
+	statement_sink: multiqueue::BroadcastFutSender<message::Statement>,
+	statement_stream: multiqueue::BroadcastFutReceiver<message::Statement>,
+	bft_message_sink: multiqueue::BroadcastFutSender<message::BftMessage>,
+	bft_message_stream: multiqueue::BroadcastFutReceiver<message::BftMessage>,
+}
+
+impl Consensus {
+	/// Create a new instance.
+	pub fn new() -> Consensus {
+		let (statement_sink, statement_stream) = multiqueue::broadcast_fut_queue(QUEUE_SIZE);
+		let (bft_sink, bft_stream) = multiqueue::broadcast_fut_queue(QUEUE_SIZE);
+		Consensus {
+			peers: HashMap::new(),
+			our_candidate: None,
+			statement_sink: statement_sink,
+			statement_stream: statement_stream,
+			bft_message_sink: bft_sink,
+			bft_message_stream: bft_stream,
+		}
+	}
+
+	/// Closes all notification streams.
+	pub fn restart(&mut self) {
+		let (statement_sink, statement_stream) = multiqueue::broadcast_fut_queue(QUEUE_SIZE);
+		let (bft_sink, bft_stream) = multiqueue::broadcast_fut_queue(QUEUE_SIZE);
+		self.statement_sink = statement_sink;
+		self.statement_stream = statement_stream;
+		self.bft_message_sink = bft_sink;
+		self.bft_message_stream = bft_stream;
+	}
+
+	/// Handle new connected peer.
+	pub fn new_peer(&mut self, _io: &mut SyncIo, _protocol: &Protocol, peer_id: PeerId, roles: &[message::Role]) {
+		if roles.iter().any(|r| *r == message::Role::Validator) {
+			trace!(target:"sync", "Registering validator {}", peer_id);
+			self.peers.insert(peer_id, PeerConsensus {
+				candidate_fetch: None,
+				candidate_available: None,
+			});
+		}
+	}
+
+	pub fn on_statement(&mut self, peer_id: PeerId, statement: message::Statement) {
+		if let Some(ref mut peer) = self.peers.get_mut(&peer_id) {
+			// TODO: validate signature?
+			match &statement.statement {
+				&message::UnsignedStatement::Candidate(ref receipt) => peer.candidate_available = Some(Hash::from(receipt.blake2_256())),
+				&message::UnsignedStatement::Available(ref hash) => peer.candidate_available = Some(*hash),
+				&message::UnsignedStatement::Valid(_) | &message::UnsignedStatement::Invalid(_) => (),
+			}
+			if let Err(e) = self.statement_sink.try_send(statement) {
+				trace!(target:"sync", "Error broadcasting statement notification: {:?}", e);
+			}
+		} else {
+			trace!(target:"sync", "Ignored statement from unregistered peer {}", peer_id);
+		}
+	}
+
+	pub fn statements(&self) -> multiqueue::BroadcastFutReceiver<message::Statement>{
+		self.statement_stream.add_stream()
+	}
+
+	pub fn on_bft_message(&mut self, peer_id: PeerId, message: message::BftMessage) {
+		if self.peers.contains_key(&peer_id) {
+			// TODO: validate signature?
```

Comment:
aren't the signatures being checked? I thought you mentioned that to me

Reply:
(it's done in `process_message` on the other side of this stream)

Reply:
Yeah, but eventually there needs to be a check here as well to avoid propagating invalid messages. Or it should be moved here.